File: ViuaBasic/Utl.cs
using System;
using System.Collections.Generic;

namespace ViuaBasic
{
  public class Utl
  {
    public static List<string> split_line(string line)
    {
      List<string> parts = new List<string>();
      string part = "";
      bool in_quote = false;
      bool out_quote = false;
      foreach (char c in line)
      {
        if (Char.IsWhiteSpace(c))
        {
          if (in_quote)
          {
            part = part + c;
          }
          else
          {
            if (out_quote)
            {
              out_quote = false;
            }
            if (part.Length > 0)
            {
              parts.Add(part);
              part = "";
            }
          }
        }
        else if (c.Equals('"'))
        {
          part = part + c;
          if (in_quote)
          {
            in_quote = false;
            out_quote = true;
          }
          else if (out_quote)
          {
            out_quote = false;
            in_quote = true;
          }
          else
          {
            in_quote = true;
          }
        }
        else
        {
          if (out_quote)
          {
            out_quote = false;
          }
          part = part + c;
        }
      }
      if (part.Length > 0)
      {
        parts.Add(part);
      }
      return parts;
    }

    public static List<string> split_separator(string line, char sep, bool trim, bool keep_sep)
    {
      List<string> parts = new List<string>();
      string part = "";
      bool in_quote = false;
      bool out_quote = false;
      foreach (char c in line)
      {
        if (c.Equals(sep))
        {
          if (in_quote)
          {
            part = part + c;
          }
          else
          {
            if (trim)
            {
              part = part.Trim();
            }
            if (part.Length > 0)
            {
              parts.Add(part);
            }
            part = "";
            if (keep_sep)
            {
              parts.Add(part + c);
            }
          }
        }
        else if (c.Equals('"'))
        {
          part = part + c;
          if (in_quote)
          {
            in_quote = false;
            out_quote = true;
          }
          else if (out_quote)
          {
            out_quote = false;
            in_quote = true;
          }
          else
          {
            in_quote = true;
          }
        }
        else
        {
          part = part + c;
          if (out_quote)
          {
            out_quote = false;
          }
        }
      }
      if (trim)
      {
        part = part.Trim();
      }
      if (part.Length > 0)
      {
        parts.Add(part);
      }
      return parts;
    }

    public static bool is_quoted(string arg)
    {
      return (arg.Length > 1) && (arg.StartsWith("\"")) && (arg.EndsWith("\""));
    }

    public static string exp_to_str(List<string> arg)
    {
      string buf = "";
      foreach (string s in arg)
      {
        buf = buf + s;
      }
      return buf;
    }

    public static List<string> exp_to_math_rpn(List<string> exp)
    {
      List<string> math = exp_to_math(exp);
      List<string> rpn = new List<string>();
      Stack<string> stack = new Stack<string>();
      bool unary = true;
      bool negate = false;
      foreach (string arg in math)
      {
        if (arg.Equals("("))
        {
          if (negate)
          {
            rpn.Add("-1");
            while (stack.Count > 0)
            {
              if (stack.Peek().Equals("*") || stack.Peek().Equals("/") || stack.Peek().Equals("%") || stack.Peek().Equals("^"))
              {
                rpn.Add(stack.Pop());
                continue;
              }
              else
              {
                break;
              }
            }
            stack.Push("*");
            negate = false;
          }
          stack.Push(arg);
          unary = true;
        }
        else if (arg.Equals(")"))
        {
          while (stack.Count > 0)
          {
            string elem = stack.Pop();
            if (elem.Equals("("))
            {
              break;
            }
            else
            {
              rpn.Add(elem);
            }
          }
          unary = false;
          negate = false;
        }
        else if (arg.Equals("+") || arg.Equals("-"))
        {
          if (unary)
          {
            if (arg.Equals("-"))
            {
              negate = !negate;
            }
          }
          else
          {
            while (stack.Count > 0)
            {
              if (stack.Peek().Equals("+") || stack.Peek().Equals("-") || stack.Peek().Equals("*") || stack.Peek().Equals("/") || stack.Peek().Equals("%") || stack.Peek().Equals("^"))
              {
                rpn.Add(stack.Pop());
                continue;
              }
              else
              {
                break;
              }
            }
            stack.Push(arg);
            unary = true;
            negate = false;
          }
        }
        else if (arg.Equals("*") || arg.Equals("/") || arg.Equals("%"))
        {
          while (stack.Count > 0)
          {
            if (stack.Peek().Equals("*") || stack.Peek().Equals("/") || stack.Peek().Equals("%") || stack.Peek().Equals("^"))
            {
              rpn.Add(stack.Pop());
              continue;
            }
            else
            {
              break;
            }
          }
          stack.Push(arg);
          unary = true;
          negate = false;
        }
        else if (arg.Equals("^"))
        {
          while (stack.Count > 0)
          {
            if (stack.Peek().Equals("^"))
            {
              rpn.Add(stack.Pop());
              continue;
            }
            else
            {
              break;
            }
          }
          stack.Push(arg);
          unary = true;
          negate = false;
        }
        else
        {
          if (negate)
          {
            rpn.Add("-" + arg);
          }
          else
          {
            rpn.Add(arg);
          }
          unary = false;
          negate = false;
        }
      }
      while (stack.Count > 0)
      {
        rpn.Add(stack.Pop());
      }
      return rpn;
    }

    public static List<string> exp_to_math(List<string> exp)
    {
      List<string> tmp1 = new List<string>();
      foreach (string elem in exp)
      {
        tmp1.AddRange(split_separator(elem, '(', true, true));
      }
      List<string> tmp2 = new List<string>();
      foreach (string elem in tmp1)
      {
        tmp2.AddRange(split_separator(elem, ')', true, true));
      }
      tmp1.Clear();
      foreach (string elem in tmp2)
      {
        tmp1.AddRange(split_separator(elem, '+', true, true));
      }
      tmp2.Clear();
      foreach (string elem in tmp1)
      {
        tmp2.AddRange(split_separator(elem, '-', true, true));
      }
      tmp1.Clear();
      foreach (string elem in tmp2)
      {
        tmp1.AddRange(split_separator(elem, '*', true, true));
      }
      tmp2.Clear();
      foreach (string elem in tmp1)
      {
        tmp2.AddRange(split_separator(elem, '/', true, true));
      }
      tmp1.Clear();
      foreach (string elem in tmp2)
      {
        tmp1.AddRange(split_separator(elem, '%', true, true));
      }
      tmp2.Clear();
      foreach (string elem in tmp1)
      {
        tmp2.AddRange(split_separator(elem, '^', true, true));
      }
      return tmp2;
    }

    public static List<string> exp_to_logic_rpn(List<string> exp)
    {
      List<string> cond = exp_to_logic(exp);
      List<string> rpn = new List<string>();
      Stack<string> stack = new Stack<string>();
      foreach (string arg in cond)
      {
        if (arg.Equals("("))
        {
          stack.Push(arg);
        }
        else if (arg.Equals(")"))
        {
          while (stack.Count > 0)
          {
            string elem = stack.Pop();
            if (elem.Equals("("))
            {
              break;
            }
            else
            {
              rpn.Add(elem);
            }
          }
        }
        else if (arg.Equals("=") || arg.Equals(">") || arg.Equals("<") || arg.Equals(">=") || arg.Equals("<=") || arg.Equals("<>"))
        {
          while (stack.Count > 0)
          {
            if (stack.Peek().Equals("=") || stack.Peek().Equals(">") || stack.Peek().Equals("<") || stack.Peek().Equals(">=") || stack.Peek().Equals("<=") || stack.Peek().Equals("<>") || stack.Peek().Equals("OR") || stack.Peek().Equals("AND") || stack.Peek().Equals("NOT"))
            {
              rpn.Add(stack.Pop());
              continue;
            }
            else
            {
              break;
            }
          }
          stack.Push(arg);
        }
        else if (arg.Equals("OR"))
        {
          while (stack.Count > 0)
          {
            if (stack.Peek().Equals("OR") || stack.Peek().Equals("AND") || stack.Peek().Equals("NOT"))
            {
              rpn.Add(stack.Pop());
              continue;
            }
            else
            {
              break;
            }
          }
          stack.Push(arg);
        }
        else if (arg.Equals("AND"))
        {
          while (stack.Count > 0)
          {
            if (stack.Peek().Equals("AND") || stack.Peek().Equals("NOT"))
            {
              rpn.Add(stack.Pop());
              continue;
            }
            else
            {
              break;
            }
          }
          stack.Push(arg);
        }
        else if (arg.Equals("NOT"))
        {
          while (stack.Count > 0)
          {
            if (stack.Peek().Equals("NOT"))
            {
              rpn.Add(stack.Pop());
              continue;
            }
            else
            {
              break;
            }
          }
          stack.Push(arg);
        }
        else
        {
          rpn.Add(arg);
        }
      }
      while (stack.Count > 0)
      {
        rpn.Add(stack.Pop());
      }
      return rpn;
    }

    public static List<string> exp_to_logic(List<string> exp)
    {
      List<string> tmp1 = new List<string>();
      foreach (string elem in exp)
      {
        tmp1.AddRange(split_separator(elem, '(', true, true));
      }
      List<string> tmp2 = new List<string>();
      foreach (string elem in tmp1)
      {
        tmp2.AddRange(split_separator(elem, ')', true, true));
      }
      tmp1.Clear();
      foreach (string elem in tmp2)
      {
        tmp1.AddRange(split_separator(elem, '<', true, true));
      }
      tmp2.Clear();
      foreach (string elem in tmp1)
      {
        tmp2.AddRange(split_separator(elem, '>', true, true));
      }
      tmp1.Clear();
      foreach (string elem in tmp2)
      {
        tmp1.AddRange(split_separator(elem, '=', true, true));
      }
      tmp2.Clear();
      int idx = 0;
      while (idx < tmp1.Count)
      {
        if ((idx + 1) < tmp1.Count)
        {
          if (tmp1[idx].Equals("<"))
          {
            if (tmp1[idx + 1].Equals("="))
            {
              tmp2.Add("<=");
              idx = idx + 2;
              continue;
            }
            if (tmp1[idx + 1].Equals(">"))
            {
              tmp2.Add("<>");
              idx = idx + 2;
              continue;
            }
          }
          if (tmp1[idx].Equals(">"))
          {
            if (tmp1[idx + 1].Equals("="))
            {
              tmp2.Add(">=");
              idx = idx + 2;
              continue;
            }
          }
        }
        if (tmp1[idx].ToUpper().Equals("OR"))
        {
          tmp2.Add("OR");
        }
        else if (tmp1[idx].ToUpper().Equals("AND"))
        {
          tmp2.Add("AND");
        }
        else if (tmp1[idx].ToUpper().Equals("NOT"))
        {
          tmp2.Add("NOT");
        }
        else
        {
          tmp2.Add(tmp1[idx]);
        }
        idx++;
      }
      idx = 0;
      tmp1.Clear();
      while (idx < tmp2.Count)
      {
        if (tmp2[idx].Equals("(") && ((idx + 1) < tmp2.Count))
        {
          int math_idx = idx + 1;
          int math_cnt = 0;
          string math = "";
          while (math_idx < tmp2.Count)
          {
            if (tmp2[math_idx].Equals("("))
            {
              math_cnt++;
              math = math + tmp2[math_idx];
              math_idx++;
              continue;
            }
            else if (tmp2[math_idx].Equals(")"))
            {
              if (math_cnt > 0)
              {
                math_cnt--;
                math = math + tmp2[math_idx];
                math_idx++;
                continue;
              }
              else
              {
                break;
              }
            }
            else if (tmp2[math_idx].Equals("=") || tmp2[math_idx].Equals(">") || tmp2[math_idx].Equals("<") || tmp2[math_idx].Equals(">=") || tmp2[math_idx].Equals("<=") || tmp2[math_idx].Equals("<>") || tmp2[math_idx].Equals("OR") || tmp2[math_idx].Equals("AND") || tmp2[math_idx].Equals("NOT"))
            {
              math = "";
              break;
            }
          }
          if (math.Length.Equals(0))
          {
            tmp1.Add(tmp2[idx]);
            idx++;
          }
          else
          {
            tmp1.Add(math);
            idx = math_idx + 1;
          }
        }
        else
        {
          tmp1.Add(tmp2[idx]);
          idx++;
        }
      }
      return tmp1;
    }

    public static List<string> take_until(int from_idx, string to_ident, List<string> parts)
    {
      List<string> res = new List<string>();
      int idx = from_idx;
      while (idx < parts.Count)
      {
        if (parts[idx].ToUpper().Equals(to_ident))
        {
          return res;
        }
        res.Add(parts[idx++]);
      }
      return res;
    }
  }
}

File: src/main/java/org/viuavm/viuabasic/Utl.java
package org.viuavm.viuabasic;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Deque;
import java.util.List;

public class Utl
{
  public static List<String> split_line(String line)
  {
    List<String> parts = new ArrayList<String>();
    String part = "";
    boolean in_quote = false;
    boolean out_quote = false;
    for (Character c : line.toCharArray())
    {
      if (Character.isWhitespace(c))
      {
        if (in_quote)
        {
          part = part + c;
        }
        else
        {
          if (out_quote)
          {
            out_quote = false;
          }
          if (part.length() > 0)
          {
            parts.add(part);
            part = "";
          }
        }
      }
      else if (c.equals('"'))
      {
        part = part + c;
        if (in_quote)
        {
          in_quote = false;
          out_quote = true;
        }
        else if (out_quote)
        {
          out_quote = false;
          in_quote = true;
        }
        else
        {
          in_quote = true;
        }
      }
      else
      {
        if (out_quote)
        {
          out_quote = false;
        }
        part = part + c;
      }
    }
    if (part.length() > 0)
    {
      parts.add(part);
    }
    return parts;
  }

  public static List<String> split_separator(String line, char sep, boolean trim, boolean keep_sep, boolean use_parentheses)
  {
    List<String> parts = new ArrayList<String>();
    String part = "";
    boolean in_quote = false;
    boolean out_quote = false;
    int parentheses = 0;
    for (Character c : line.toCharArray())
    {
      if (c.equals(sep))
      {
        if (use_parentheses && (parentheses > 0))
        {
          part = part + c;
        }
        else if (in_quote)
        {
          part = part + c;
        }
        else
        {
          if (trim)
          {
            part = part.trim();
          }
          if (part.length() > 0)
          {
            parts.add(part);
          }
          part = "";
          if (keep_sep)
          {
            parts.add(part + c);
          }
        }
      }
      else if (c.equals('"'))
      {
        part = part + c;
        if (in_quote)
        {
          in_quote = false;
          out_quote = true;
        }
        else if (out_quote)
        {
          out_quote = false;
          in_quote = true;
        }
        else
        {
          in_quote = true;
        }
      }
      else
      {
        part = part + c;
        if (out_quote)
        {
          out_quote = false;
        }
        if (use_parentheses)
        {
          if (c.equals('('))
          {
            parentheses++;
          }
          if (c.equals(')'))
          {
            parentheses--;
          }
        }
      }
    }
    if (trim)
    {
      part = part.trim();
    }
    if (part.length() > 0)
    {
      parts.add(part);
    }
    return parts;
  }

  public static List<String> list_split_separator(List<String> lines, char sep, boolean trim, boolean keep_sep, boolean use_parentheses)
  {
    List<String> result = new ArrayList<String>();
    for (String line : lines)
    {
      result.addAll(split_separator(line, sep, trim, keep_sep, use_parentheses));
    }
    return result;
  }

  public static boolean is_quoted(String arg)
  {
    return (arg.length() > 1) && (arg.startsWith("\"")) && (arg.endsWith("\""));
  }

  public static String escape_quotes(String arg)
  {
    String result = arg;
    if (arg.length() > 3)
    {
      result = arg.substring(1, arg.length() - 1);
      result = result.replace("\"\"", "\\\"");
      result = "\"" + result + "\"";
    }
    return result;
  }

  public static String exp_to_str(List<String> arg)
  {
    String buf = "";
    for (String s : arg)
    {
      buf = buf + s;
    }
    return buf;
  }

  public static List<String> exp_to_math_rpn(List<String> exp, Variables vars)
  {
    List<String> math = exp_to_math(exp);
    List<String> rpn = new ArrayList<String>();
    Deque<String> stack = new ArrayDeque<String>();
    boolean unary = true;
    boolean negate = false;
    for (String arg : math)
    {
      if (arg.equals("("))
      {
        if (negate)
        {
          rpn.add("-1");
          while (stack.size() > 0)
          {
            if (stack.peek().equals("*") || stack.peek().equals("/") || stack.peek().equals("%") || stack.peek().equals("^"))
            {
              rpn.add(stack.pop());
              continue;
            }
            else
            {
              break;
            }
          }
          stack.push("*");
          negate = false;
        }
        stack.push(arg);
        unary = true;
      }
      else if (arg.equals(")"))
      {
        while (stack.size() > 0)
        {
          String elem = stack.pop();
          if (elem.equals("("))
          {
            if (stack.size() > 0)
            {
              elem = stack.peek();
              if (elem.equals("ABS") || elem.equals("EXP") || elem.equals("LOG") || vars.is_array(elem))
              {
                rpn.add(stack.pop());
              }
            }
            break;
          }
          else
          {
            rpn.add(elem);
          }
        }
        unary = false;
        negate = false;
      }
      else if (arg.equals(","))
      {
        while (stack.size() > 0)
        {
          if (stack.peek().equals(",") || stack.peek().equals("+") || stack.peek().equals("-") || stack.peek().equals("*") || stack.peek().equals("/") || stack.peek().equals("%") || stack.peek().equals("^"))
          {
            rpn.add(stack.pop());
            continue;
          }
          else
          {
            break;
          }
        }
        stack.push(arg);
        unary = true;
        negate = false;
      }
      else if (arg.equals("+") || arg.equals("-"))
      {
        if (unary)
        {
          if (arg.equals("-"))
          {
            negate = !negate;
          }
        }
        else
        {
          while (stack.size() > 0)
          {
            if (stack.peek().equals("+") || stack.peek().equals("-") || stack.peek().equals("*") || stack.peek().equals("/") || stack.peek().equals("%") || stack.peek().equals("^"))
            {
              rpn.add(stack.pop());
              continue;
            }
            else
            {
              break;
            }
          }
          stack.push(arg);
          unary = true;
          negate = false;
        }
      }
      else if (arg.equals("*") || arg.equals("/") || arg.equals("%"))
      {
        while (stack.size() > 0)
        {
          if (stack.peek().equals("*") || stack.peek().equals("/") || stack.peek().equals("%") || stack.peek().equals("^"))
          {
            rpn.add(stack.pop());
            continue;
          }
          else
          {
            break;
          }
        }
        stack.push(arg);
        unary = true;
        negate = false;
      }
      else if (arg.equals("^"))
      {
        while (stack.size() > 0)
        {
          if (stack.peek().equals("^"))
          {
            rpn.add(stack.pop());
            continue;
          }
          else
          {
            break;
          }
        }
        stack.push(arg);
        unary = true;
        negate = false;
      }
      else if (arg.toUpperCase().equals("ABS") || arg.toUpperCase().equals("EXP") || arg.toUpperCase().equals("LOG") || vars.is_array(arg.toUpperCase()))
      {
        stack.push(arg.toUpperCase());
        unary = false;
        negate = false;
      }
      else
      {
        if (negate)
        {
          rpn.add("-" + arg);
        }
        else
        {
          rpn.add(arg);
        }
        unary = false;
        negate = false;
      }
    }
    while (stack.size() > 0)
    {
      rpn.add(stack.pop());
    }
    return rpn;
  }

  public static List<String> exp_to_math(List<String> exp)
  {
    List<String> tmp = list_split_separator(exp, '(', true, true, false);
    tmp = list_split_separator(tmp, ')', true, true, false);
    tmp = list_split_separator(tmp, ',', true, true, false);
    tmp = list_split_separator(tmp, '+', true, true, false);
    tmp = list_split_separator(tmp, '-', true, true, false);
    tmp = list_split_separator(tmp, '*', true, true, false);
    tmp = list_split_separator(tmp, '/', true, true, false);
    tmp = list_split_separator(tmp, '%', true, true, false);
    tmp = list_split_separator(tmp, '^', true, true, false);
    return tmp;
  }

  public static List<String> exp_to_logic_rpn(List<String> exp, Variables vars)
  {
    List<String> cond = exp_to_logic(exp, vars);
    List<String> rpn = new ArrayList<String>();
    Deque<String> stack = new ArrayDeque<String>();
    for (String arg : cond)
    {
      if (arg.equals("("))
      {
        stack.push(arg);
      }
      else if (arg.equals(")"))
      {
        while (stack.size() > 0)
        {
          String elem = stack.pop();
          if (elem.equals("("))
          {
            break;
          }
          else
          {
            rpn.add(elem);
          }
        }
      }
      else if (arg.equals("=") || arg.equals(">") || arg.equals("<") || arg.equals(">=") || arg.equals("<=") || arg.equals("<>"))
      {
        while (stack.size() > 0)
        {
          if (stack.peek().equals("=") || stack.peek().equals(">") || stack.peek().equals("<") || stack.peek().equals(">=") || stack.peek().equals("<=") || stack.peek().equals("<>") || stack.peek().equals("OR") || stack.peek().equals("AND") || stack.peek().equals("NOT"))
          {
            rpn.add(stack.pop());
            continue;
          }
          else
          {
            break;
          }
        }
        stack.push(arg);
      }
      else if (arg.equals("OR"))
      {
        while (stack.size() > 0)
        {
          if (stack.peek().equals("OR") || stack.peek().equals("AND") || stack.peek().equals("NOT"))
          {
            rpn.add(stack.pop());
            continue;
          }
          else
          {
            break;
          }
        }
        stack.push(arg);
      }
      else if (arg.equals("AND"))
      {
        while (stack.size() > 0)
        {
          if (stack.peek().equals("AND") || stack.peek().equals("NOT"))
          {
            rpn.add(stack.pop());
            continue;
          }
          else
          {
            break;
          }
        }
        stack.push(arg);
      }
      else if (arg.equals("NOT"))
      {
        while (stack.size() > 0)
        {
          if (stack.peek().equals("NOT"))
          {
            rpn.add(stack.pop());
            continue;
          }
          else
          {
            break;
          }
        }
        stack.push(arg);
      }
      else
      {
        rpn.add(arg);
      }
    }
    while (stack.size() > 0)
    {
      rpn.add(stack.pop());
    }
    return rpn;
  }

  public static List<String> exp_to_logic(List<String> exp, Variables vars)
  {
    List<String> tmp1 = list_split_separator(exp, '(', true, true, false);
    tmp1 = list_split_separator(tmp1, ')', true, true, false);
    tmp1 = list_split_separator(tmp1, '<', true, true, false);
    tmp1 = list_split_separator(tmp1, '>', true, true, false);
    tmp1 = list_split_separator(tmp1, '=', true, true, false);
    List<String> tmp2 = new ArrayList<String>();
    int idx = 0;
    while (idx < tmp1.size())
    {
      if ((idx + 1) < tmp1.size())
      {
        if (tmp1.get(idx).equals("<"))
        {
          if (tmp1.get(idx + 1).equals("="))
          {
            tmp2.add("<=");
            idx = idx + 2;
            continue;
          }
          if (tmp1.get(idx + 1).equals(">"))
          {
            tmp2.add("<>");
            idx = idx + 2;
            continue;
          }
        }
        if (tmp1.get(idx).equals(">"))
        {
          if (tmp1.get(idx + 1).equals("="))
          {
            tmp2.add(">=");
            idx = idx + 2;
            continue;
          }
        }
      }
      if (tmp1.get(idx).toUpperCase().equals("OR"))
      {
        tmp2.add("OR");
      }
      else if (tmp1.get(idx).toUpperCase().equals("AND"))
      {
        tmp2.add("AND");
      }
      else if (tmp1.get(idx).toUpperCase().equals("NOT"))
      {
        tmp2.add("NOT");
      }
      else
      {
        tmp2.add(tmp1.get(idx));
      }
      idx++;
    }
    idx = 0;
    tmp1.clear();
    while (idx < tmp2.size())
    {
      if (tmp2.get(idx).equals("(") && ((idx + 1) < tmp2.size()))
      {
        int math_idx = idx + 1;
        int math_cnt = 0;
        String math = "";
        while (math_idx < tmp2.size())
        {
          if (tmp2.get(math_idx).equals("("))
          {
            math_cnt++;
            math = math + tmp2.get(math_idx++);
            continue;
          }
          else if (tmp2.get(math_idx).equals(")"))
          {
            if (math_cnt > 0)
            {
              math_cnt--;
              math = math + tmp2.get(math_idx++);
              continue;
            }
            else
            {
              break;
            }
          }
          else if (tmp2.get(math_idx).equals("=") || tmp2.get(math_idx).equals(">") || tmp2.get(math_idx).equals("<") || tmp2.get(math_idx).equals(">=") || tmp2.get(math_idx).equals("<=") || tmp2.get(math_idx).equals("<>") || tmp2.get(math_idx).equals("OR") || tmp2.get(math_idx).equals("AND") || tmp2.get(math_idx).equals("NOT"))
          {
            math = "";
            break;
          }
          math = math + tmp2.get(math_idx++);
        }
        if (math.isEmpty())
        {
          tmp1.add(tmp2.get(idx));
          idx++;
        }
        else
        {
          if ((tmp1.size() > 0) && (vars.is_array(tmp1.get(tmp1.size() - 1).toUpperCase())))
          {
            tmp1.set(tmp1.size() - 1, tmp1.get(tmp1.size() - 1) + "(" + math + ")");
          }
          else
          {
            tmp1.add(math);
          }
          idx = math_idx + 1;
        }
      }
      else
      {
        tmp1.add(tmp2.get(idx));
        idx++;
      }
    }
    return tmp1;
  }

  public static List<String> take_until(int from_idx, String to_ident, List<String> parts)
  {
    List<String> result = new ArrayList<String>();
    int idx = from_idx;
    while (idx < parts.size())
    {
      if (parts.get(idx).toUpperCase().equals(to_ident))
      {
        return result;
      }
      result.add(parts.get(idx++));
    }
    return result;
  }
}
